docs/guides/megascale_hang_playbook.md

```diff
@@ -0,0 +1,112 @@
+
+The following error message indicates that there is either a faulty TPU, or the workers are stalling on the input pipeline.
+
+```
+Megascale detects a hang that is likely caused by bad TPU chips on the following hosts. Please remove the hosts from the fleet and restart the workload. If problem persists please contact Megascale XLA team.
+The host that have bad TPUs are: <host>
+Full error digest:
+  Potential cause: Bad TPU chips
+  Potential culprit workers: <host>
+  TPU states:
+    Module: jit_loss_and_grad
+    Fingerprint: <fingerprint>
+    Launch ID: <id>
+      <tag>:<pc>(<hlo>): <host>
+```
+
+If the TPU listed in the log shows a non-zero program counter, it is very likely that the TPU is faulty and causing the hang. Restart the job and check if the same TPU shows up as an outlier. If the TPU repeatedly shows up as an outlier, send the information to Google so we can remove the host from the fleet.
+
+If the logged TPU shows a program counter of 0, it is likely that the TPU is waiting on input. We can attempt to confirm the worker is hung during the input pipeline using the stack trace library found in the [cloud-tpu-diagnostics package](https://pypi.org/project/cloud-tpu-diagnostics/).
+
+XPK users should follow the [XPK-specific instructions](https://github.com/AI-Hypercomputer/xpk?tab=readme-ov-file#collect-stack-traces) to emit stack traces. Note the `--deploy-stacktrace-sidecar` flag when running the XPK workload command.
+
+Customers can then query Cloud Logging for the stack trace logs from the outlier TPU. The stack trace log will help users determine where in the Python code the program was during the hang.
+
+### 3. Unknown Error
+
+The following error indicates the MXLA runtime cannot determine the cause of the hangs. Send Google an HLO dump (see how [here](https://openxla.org/xla/hlo_dumps)) for further debugging.
+
+```
+Megascale detects a hang but cannot determine the root cause. Please inspect the full digest below.
+```
+
+### 4. Checkpoint File Sizes Too Small
+
+If your checkpoint file sizes are too small and frequent, there is a risk of GCS issues. We recommend using [Orbax checkpointing](https://orbax.readthedocs.io/en/latest/index.html) which writes 200MB files by default.
+
+### 5. Training code uses MXLA to synchronize for a I/O operation
+
+Training code might use [MXLA collectives](https://openxla.org/xla) as a global barrier to sync across devices.
+
+In case one of the hosts is taking more than the time set in `--megascale_graph_hang_threshold` (default is 5 minutes) then MXLA timeout will expire. Depending on the setting, MXLA will either block the training or will crash and restart it. If the latter, then a series of restarts will occur until the I/O operation completes within the timeout period.
+
+For example, use
+
+```python
+jax._src.distributed.global_state.client.wait_at_barrier("<barrier_id>")
+multihost_utils.sync_global_devices(...)
+```
+
+for worker synchronization.
```

Comment:
How about something like this instead?

As an example, this could occur when using an all-reduce to synchronize at the end of checkpoint loading. One possible solution would be to switch this collective barrier to use the JAX distributed `wait_at_barrier` API instead:

```
jax._src.distributed.global_state.client.wait_at_barrier(...)
```